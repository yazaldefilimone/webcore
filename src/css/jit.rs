use crate::css;

// Implement a JIT compiler for CSS baded on WebKit's CSS JIT compiler.
// https://www.webkit.org/blog/3271/webkit-css-selector-jit-compiler/
//
//

pub struct CssJitCompiler {
  pub stylesheet: css::StyleSheet,
}

impl CssJitCompiler {
  pub fn new(stylesheet: css::StyleSheet) -> Self {
    CssJitCompiler { stylesheet }
  }
}
