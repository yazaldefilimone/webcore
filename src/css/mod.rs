#![allow(dead_code, unused_imports)]
mod jit;
mod node;
mod parser;
mod tree;
pub use node::*;
pub use parser::CSSParser;
pub use tree::*;
