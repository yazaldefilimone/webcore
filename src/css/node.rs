#![allow(dead_code)]
#[derive(Debug, Clone, serde::Serialize)]
pub struct StyleSheet {
  pub rules: Vec<Rule>,
}
#[derive(Debug, Clone, serde::Serialize)]
pub struct Rule {
  pub selectors: Vec<Selector>,
  pub declarations: Vec<Declaration>,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct Declaration {
  pub name: String,
  pub value: DeclarationValue,
}

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
pub enum DeclarationValue {
  Keyword(String),
  Length(f32, UnitValue),
  ColorValue(ColorValue),
}

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
pub enum ColorValue {
  HexColorValue(String),
  RBGColorValue(u8, u8, u8, u8),
  HSLColorValue(u8, u8, u8, u8),
}

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
pub enum UnitValue {
  Px,
  Em,
  Rem,
  Percent,
}

#[derive(Debug, Clone, serde::Serialize)]
pub enum Selector {
  // .selector
  Simple(SimpleSelector),
}
#[derive(Debug, Clone, serde::Serialize)]
pub struct SimpleSelector {
  pub tag_name: Option<String>,
  pub id: Option<String>,
  pub class: Vec<String>,
}

impl SimpleSelector {
  pub fn new() -> Self {
    SimpleSelector { tag_name: None, id: None, class: Vec::new() }
  }
}

pub type Specificity = (usize, usize, usize);

impl Selector {
  pub fn new() -> Self {
    Selector::Simple(SimpleSelector::new())
  }
  pub fn specificity(&self) -> Specificity {
    // http://www.w3.org/TR/selectors/#specificity
    let Selector::Simple(SimpleSelector { tag_name, id, class }) = self;
    return (tag_name.iter().count(), id.iter().count(), class.len());
  }
}

impl DeclarationValue {
  pub fn to_px(&self) -> f32 {
    match self {
      DeclarationValue::Length(px, _) => *px,
      // TODO: support other unit
      _ => 0.0,
    }
  }
}
