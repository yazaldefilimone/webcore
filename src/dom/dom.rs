#![allow(dead_code)]
use std::collections::{BTreeMap, HashMap, HashSet};

use serde::{ser::SerializeStruct, Serialize, Serializer};

#[derive(Debug, Clone, serde::Serialize)]
pub struct Node {
  pub children: Vec<Node>,
  pub node_type: NodeType,
}
#[derive(Debug, Clone, serde::Serialize)]
pub enum NodeType {
  Text(String),
  Element(Element),
}

pub type AtributeMapType = HashMap<String, String>;

#[derive(Debug, Clone)]
pub struct Element {
  pub tag_name: String,
  pub atributes: AtributeMapType,
}

// This is a better implementation of Serialize for HashMap? (we use it in snapshot tests)
impl Serialize for Element {
  fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
  where
    S: Serializer,
  {
    let ordered_attributes: BTreeMap<_, _> = self.atributes.iter().collect();
    let mut state = serializer.serialize_struct("Element", 2)?;
    state.serialize_field("tag_name", &self.tag_name)?;
    state.serialize_field("atributes", &ordered_attributes)?;
    state.end()
  }
}

// Doctype
//
#[derive(Debug, Clone, serde::Serialize)]
pub struct Doctype {
  pub name: String,
  pub public_id: Option<String>,
  pub system_id: Option<String>,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct HtmlRoot {
  pub doctype: Option<Doctype>,
  pub children: Vec<Node>,
}

pub fn create_text(text: String) -> Node {
  Node { children: Vec::new(), node_type: NodeType::Text(text) }
}
pub fn create_element(tag_name: String, atributes: AtributeMapType, children: Vec<Node>) -> Node {
  Node { children, node_type: NodeType::Element(Element { tag_name, atributes }) }
}

impl Element {
  pub fn id(&self) -> String {
    self.atributes.get("id").unwrap().to_string()
  }

  pub fn classes(&self) -> HashSet<&str> {
    let mut classes = HashSet::new();
    for class in self.atributes.get("class").unwrap().split(" ") {
      classes.insert(class);
    }
    classes
  }

  pub fn has_class(&self, class_name: &str) -> bool {
    self.classes().contains(class_name)
  }
}

impl Node {
  pub fn tag_name(&self) -> String {
    match &self.node_type {
      NodeType::Text(_) => "".to_string(),
      NodeType::Element(element) => element.tag_name.clone(),
    }
  }
  pub fn text(&self) -> Option<String> {
    match &self.node_type {
      NodeType::Text(text) => Some(text.clone()),
      _ => None,
    }
  }

  pub fn get_children(&self) -> Vec<Node> {
    self.children.clone()
  }
}
