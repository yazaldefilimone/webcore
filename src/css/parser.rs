use crate::css;
use crate::utils::validation::is_tag_char;
// ==============================
// https://www.w3.org/TR/css-syntax-3/#tokenizing-and-parsing
//

struct Parser {
  input: String,
  cursor: usize,
}

impl Parser {
  pub fn new(input: String) -> Self {
    Parser { input, cursor: 0 }
  }
}

pub fn parse(input: String) -> css::StyleSheet {
  let mut parser = Parser::new(input);
  return css::StyleSheet { rules: parser.parse_rules() };
}

impl Parser {
  fn parse_rules(&mut self) -> Vec<css::Rule> {
    let mut rules = Vec::new();
    while !self.is_end() {
      rules.push(self.parse_rule());
    }
    rules
  }
  // Parse a rule set: `<selectors> { <declarations> }`.
  fn parse_rule(&mut self) -> css::Rule {
    css::Rule { selectors: self.parse_selectors(), declarations: self.parse_declarations() }
  }
  // simple selector, e.g.: `type#id.class1.class2.class3`
  fn parse_simple_selector(&mut self) -> css::SimpleSelector {
    let mut simple_selector = css::SimpleSelector::new();
    while !self.is_end() {
      match self.peek_one() {
        '.' => {
          self.consume_expect(".");
          simple_selector.class.push(self.parse_identifier());
        }
        '#' => {
          self.consume_expect("#");
          simple_selector.id = Some(self.parse_identifier());
        }
        character if is_tag_char(character) => {
          simple_selector.tag_name = Some(self.parse_identifier());
        }
        _ => break,
      }
    }
    simple_selector
  }

  fn parse_selectors(&mut self) -> Vec<css::Selector> {
    let mut selectors = Vec::new();
    while !self.is_end() {
      selectors.push(self.parse_selector());
      self.skip_whitespace();
      match self.peek_one() {
        ',' => {
          self.consume_expect(",");
          self.skip_whitespace();
        }
        '}' => break,
        character => {
          panic!("Unexpected character '{}' in selector list", character);
        }
      }
    }
    // return selectors with highest specificity first, for use in matching.
    selectors.sort_by(|left, right| right.specificity().cmp(&left.specificity()));
    selectors
  }

  /// Parse a list of declarations enclosed in `{ ... }`.
  fn parse_declarations(&mut self) -> Vec<css::Declaration> {
    self.consume_expect("{");
    let mut declarations = Vec::new();
    while !self.is_end() && !self.starts_with("}") {
      self.skip_whitespace();
      declarations.push(self.parse_declaration());
      self.skip_whitespace();
    }
    self.consume_expect("}");
    declarations
  }
  /// Parse one `<property>: <value>;` declaration.
  fn parse_declaration(&mut self) -> css::Declaration {
    let property_name = self.parse_identifier();
    self.consume_expect(":");
    self.skip_whitespace();
    let value = self.parse_value();
    self.consume_expect(";");
    css::Declaration { name: property_name, value }
  }

  fn parse_value(&mut self) -> css::DeclarationValue {
    let value = self.parse_identifier();
    css::DeclarationValue::Keyword(value)
  }

  fn parse_value_length(&mut self) -> css::DeclarationValue {
    css::DeclarationValue::Length(self.parse_float(), self.parse_unit())
  }

  fn parse_unit(&mut self) -> css::UnitValue {
    let unit = self.parse_identifier();
    match &*unit.to_ascii_lowercase() {
      "px" => css::UnitValue::Px,
      "em" => css::UnitValue::Em,
      "rem" => css::UnitValue::Rem,
      _ => panic!("Unknown unit '{}'", unit),
    }
  }

  fn parse_float(&mut self) -> f32 {
    let float = self.consume_while(|character| match character {
      '0'..='9' | '.' => true,
      _ => false,
    });
    float.parse().unwrap()
  }

  // todo: support other color formats, e.g. hex, rgba, hsla...
  fn parse_color(&mut self) -> css::DeclarationValue {
    self.consume_expect("#");
    let r = self.parse_hex_pair();
    let g = self.parse_hex_pair();
    let b = self.parse_hex_pair();
    css::DeclarationValue::Color(css::ColorValue { r, g, b, a: 255 })
  }

  /// Parse two hexadecimal digits.
  fn parse_hex_pair(&mut self) -> u8 {
    let value = &self.input[self.cursor..self.cursor + 2];
    self.cursor += 2;
    u8::from_str_radix(value, 16).unwrap()
  }
  fn parse_length(&mut self) -> f32 {
    let value = self.parse_number();
    value.parse::<f32>().unwrap()
  }

  fn parse_number(&mut self) -> String {
    let mut number = String::new();
    while !self.is_end() {
      match self.peek_one() {
        '0'..='9' => number.push(self.consume()),
        _ => break,
      }
    }
    number
  }

  // todo: support more complex selectors.
  fn parse_selector(&mut self) -> css::Selector {
    self.parse_simple_selector();
    css::Selector::Simple(self.parse_simple_selector())
  }

  fn parse_identifier(&mut self) -> String {
    return self.consume_while(is_tag_char);
  }

  fn skip_whitespace(&mut self) {
    self.consume_while(char::is_whitespace);
  }

  fn consume_while<F: Fn(char) -> bool>(&mut self, test: F) -> String {
    let start_cursor = self.cursor;
    while !self.is_end() && test(self.peek_one()) {
      self.advance_one()
    }
    return self.input[start_cursor..self.cursor].to_string();
  }

  fn is_end(&self) -> bool {
    self.cursor >= self.input.len()
  }

  fn peek_one(&self) -> char {
    self.input[self.cursor..].chars().next().unwrap()
  }

  fn peek_many(&self, count: usize) -> String {
    self.input[self.cursor..].chars().take(count).collect()
  }

  fn starts_with(&self, s: &str) -> bool {
    self.input[self.cursor..].starts_with(s)
  }

  fn advance_one(&mut self) {
    self.cursor += 1;
  }

  fn advance_many(&mut self, count: usize) {
    self.cursor += count;
  }

  fn consume(&mut self) -> char {
    let mut iter = self.input[self.cursor..].char_indices();
    let (_, cur_char) = iter.next().unwrap();
    let (next_cursor, _) = iter.next().unwrap_or((1, ' '));
    self.cursor += next_cursor;
    cur_char
  }

  fn consume_expect(&mut self, text: &str) {
    if &self.peek_many(text.len()) == text {
      self.advance_many(text.len());
    } else {
      panic!("Expected '{}' but got '{}'", text, &self.peek_many(text.len()));
    }
  }
}
