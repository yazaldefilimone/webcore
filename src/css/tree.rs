use crate::css::{DeclarationValue, Rule, Selector, SimpleSelector, Specificity, StyleSheet};
use crate::dom::{Element, Node, NodeType};
use std::collections::HashMap;
// ==================================
// https://wiki.mozilla.org/Gecko:Key_Gecko_Structures_And_Invariants
// map from CSS property names to values.
// todo: make eayse to  suport of compound selectors, e.g. `div.test p` or `div.test > p`.
//
type PropertyMap = HashMap<String, DeclarationValue>;

// A node with associated style data.
#[derive(Debug, Clone)]
pub struct StyledNode<'a> {
  pub node: &'a Node, // pointer to a DOM node
  pub specified_values: PropertyMap,
  pub children: Vec<StyledNode<'a>>,
}

// http://www.w3.org/TR/CSS2/selector.html#pattern-matching
fn matches(elem: &Element, selector: &Selector) -> bool {
  match *selector {
    Selector::Simple(ref simple_selector) => matches_simple_selector(elem, simple_selector),
  }
}

fn matches_simple_selector(elem: &Element, selector: &SimpleSelector) -> bool {
  // Check type selector
  if selector.tag_name.iter().any(|name| elem.tag_name != *name) {
    return false;
  }

  // Check ID selector
  if selector.id.iter().any(|id| elem.id() != *id) {
    return false;
  }

  // Check class selectors
  let elem_classes = elem.classes();
  if selector.class.iter().any(|class| !elem_classes.contains(&**class)) {
    return false;
  }
  // We didn't find any non-matching selector components.
  return true;
}

type MatchedRule<'a> = (Specificity, &'a Rule);

// If `rule` matches `elem`, return a `MatchedRule`. Otherwise return `None`.
fn match_rule<'a>(elem: &Element, rule: &'a Rule) -> Option<MatchedRule<'a>> {
  // Find the first (highest-specificity) matching selector.
  let mut selector_vector = rule.selectors.iter();
  let selector = selector_vector.find(|selector| matches(elem, selector));
  selector.map(|selector| (selector.specificity(), rule))
}

// Find all CSS rules that match the given element.
fn matching_rules<'a>(elem: &Element, styleheet: &'a StyleSheet) -> Vec<MatchedRule<'a>> {
  let rule_vector = styleheet.rules.iter();
  rule_vector.filter_map(|rule| match_rule(elem, rule)).collect()
}

// Apply styles to a single element, returning the specified values.
fn specified_values(elem: &Element, styleheet: &StyleSheet) -> PropertyMap {
  let mut values_map = HashMap::new();
  let mut rules_matched = matching_rules(elem, styleheet);
  // Go through the rules from lowest to highest specificity.
  rules_matched.sort_by(|&(right, _), &(left, _)| left.cmp(&right));
  for (_, rule) in rules_matched {
    for declaration in &rule.declarations {
      values_map.insert(declaration.name.clone(), declaration.value.clone());
    }
  }
  return values_map;
}

// Apply a stylesheet to an entire DOM tree, returning a StyledNode tree.
pub fn style_tree<'a>(root: &'a Node, stylesheet: &'a StyleSheet) -> StyledNode<'a> {
  let specified_values = match root.node_type {
    NodeType::Element(ref elem) => specified_values(elem, stylesheet),
    NodeType::Text(_) => HashMap::new(),
  };

  let children = root.children.iter().map(|child| style_tree(child, stylesheet));

  StyledNode { node: root, specified_values, children: children.collect() }
}
