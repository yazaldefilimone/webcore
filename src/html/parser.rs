#![allow(dead_code)]
use crate::dom;
use crate::utils::validation::is_tag_char;

pub struct HTMLParser {
  input: String,
  cursor: usize,
}

impl HTMLParser {
  pub fn new(input: String) -> Self {
    HTMLParser { input: input.trim().to_string(), cursor: 0 }
  }

  pub fn parse_root_children(&mut self) -> Vec<dom::Node> {
    let nodes = self.parse_nodes();
    return nodes;
  }
  pub fn parse_root(&mut self) -> dom::HtmlRoot {
    let doctype = self.parse_doctype();
    let children = self.parse_root_children();
    return dom::HtmlRoot { doctype, children };
  }

  pub fn parse_doctype(&mut self) -> Option<dom::Doctype> {
    if self.peek_many(2) != "<!" {
      return None;
    }
    self.consume_expect("<!");
    self.skip_whitespace();
    // consume the name of the doctype
    self.consume_while(is_tag_char);
    self.skip_whitespace();
    let name = self.consume_while(is_tag_char);
    self.skip_whitespace();
    let public_id = self.parse_public_id();
    self.skip_whitespace();
    let system_id = self.parse_system_id();
    self.consume_expect(">");
    Some(dom::Doctype { name, public_id, system_id })
  }

  fn parse_public_id(&mut self) -> Option<String> {
    if self.peek_many(6) != "PUBLIC" {
      return None;
    }
    self.consume_expect("PUBLIC");
    self.consume_expect("\"");
    let public_id = self.consume_while(|character| character != '"');
    self.consume_expect("\"");
    Some(public_id)
  }

  fn parse_system_id(&mut self) -> Option<String> {
    if self.peek_many(6) != "SYSTEM" {
      return None;
    }
    self.consume_expect("SYSTEM");
    self.consume_expect("\"");
    let system_id = self.consume_while(|character| character != '"');
    self.consume_expect("\"");
    Some(system_id)
  }

  fn parse_nodes(&mut self) -> Vec<dom::Node> {
    self.skip_trivial();
    let mut nodes = Vec::new();
    while !self.is_end() && !self.starts_with("</") {
      nodes.push(self.parse_node());
      self.skip_trivial();
    }
    nodes
  }

  fn parse_node(&mut self) -> dom::Node {
    match self.peek_one() {
      '<' => self.parse_element(),
      _ => self.parse_text(),
    }
  }

  fn parse_element(&mut self) -> dom::Node {
    self.consume_expect("<");
    let tag_name = self.parse_tag_name();
    let attributes = self.parse_attributes();
    // suport self-closing tag
    if self.peek_many(2) == "/>" {
      self.consume_expect("/>");
      return dom::create_element(tag_name, attributes, Vec::new());
    }
    self.consume_expect(">");
    let children = self.parse_nodes();

    self.consume_expect("<");
    self.consume_expect("/");
    self.consume_expect(&tag_name);
    self.consume_expect(">");
    dom::create_element(tag_name, attributes, children)
  }

  fn parse_text(&mut self) -> dom::Node {
    let text = self.consume_while(|character| character != '<');
    dom::create_text(text)
  }

  fn parse_tag_name(&mut self) -> String {
    self.consume_while(is_tag_char)
  }

  fn parse_attribute(&mut self) -> (String, String) {
    let name = self.parse_tag_name();
    self.consume_expect("=");
    let value = self.parse_attribute_value();
    (name, value)
  }

  fn parse_attribute_value(&mut self) -> String {
    let open_quote = self.consume();
    assert!(open_quote == '"' || open_quote == '\'');
    let value = self.consume_while(|c| c != open_quote);
    self.consume_expect(open_quote.to_string().as_str());
    value
  }

  fn parse_attributes(&mut self) -> dom::AtributeMapType {
    let mut attributes = dom::AtributeMapType::new();
    self.skip_whitespace();
    while self.peek_one() != '>' && self.peek_one() != '/' {
      let (name, value) = self.parse_attribute();
      attributes.insert(name, value);
      self.skip_whitespace();
    }
    attributes
  }

  fn skip_whitespace(&mut self) {
    self.consume_while(char::is_whitespace);
  }

  fn skip_trivial(&mut self) {
    self.skip_whitespace();
  }

  fn peek_one(&self) -> char {
    self.input[self.cursor..].chars().next().unwrap()
  }

  fn peek_many(&self, count: usize) -> String {
    self.input[self.cursor..].chars().take(count).collect()
  }

  fn starts_with(&self, s: &str) -> bool {
    self.input[self.cursor..].starts_with(s)
  }

  fn is_end(&self) -> bool {
    self.cursor >= self.input.len()
  }

  fn advance_one(&mut self) {
    self.cursor += 1;
  }

  fn advance_many(&mut self, count: usize) {
    self.cursor += count;
  }

  fn consume(&mut self) -> char {
    let mut iter = self.input[self.cursor..].char_indices();
    let (_, cur_char) = iter.next().unwrap();
    let (next_cursor, _) = iter.next().unwrap_or((1, ' '));
    self.cursor += next_cursor;
    cur_char
  }

  fn consume_expect(&mut self, text: &str) {
    if &self.peek_many(text.len()) == text {
      self.advance_many(text.len());
    } else {
      panic!("Expected '{}' but got '{}'", text, &self.peek_many(text.len()));
    }
  }

  fn consume_while(&mut self, mut test: impl FnMut(char) -> bool) -> String {
    let start_cursor = self.cursor;
    while !self.is_end() && test(self.peek_one()) {
      self.advance_one();
    }
    self.input[start_cursor..self.cursor].to_string()
  }
}
