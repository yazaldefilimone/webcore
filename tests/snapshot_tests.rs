#![allow(dead_code, unused_imports)]

use glob::glob;
use insta::assert_ron_snapshot;
use std::fs;
use std::path::Path;
use webcore::css::{self, CSSParser};
use webcore::dom;
use webcore::html::HTMLParser;

fn read_test_files_with_pattern(pattern: &str) -> Vec<(String, String)> {
  let mut patterns = Vec::new();
  let glob_pattern = glob(pattern).expect("Failed to read glob pattern");
  for entry in glob_pattern {
    let path = entry.expect("Failed to read file");
    let file_name = path.file_name().unwrap().to_string_lossy().to_string();
    let content = fs::read_to_string(path).expect("Failed to read file");
    patterns.push((file_name.to_owned(), content));
  }
  patterns
}
fn format_file_name_with_module(file_name: &str, module: &str, ext: &str) -> String {
  let file_name = format!("{}_{}", module, file_name).replace(ext, "");
  return file_name;
}
fn setings_snapshot() -> insta::Settings {
  let mut settings = insta::Settings::clone_current();
  settings.set_prepend_module_to_snapshot(false);
  settings.set_omit_expression(true);
  settings
}

fn create_syle_sheet_parser(source_code: &str) -> css::StyleSheet {
  let mut parser = CSSParser::new(source_code.to_string());
  return parser.parse_syle_sheet();
}
fn create_html_parser(source_code: &str) -> dom::HtmlRoot {
  let mut parser = HTMLParser::new(source_code.to_string());
  return parser.parse_root();
}

#[test]
fn test_css_parser_snapshot() {
  let test_files = read_test_files_with_pattern("tests/golden_tests/css/*.css");
  let settings = setings_snapshot();
  settings.bind(|| {
    for (file_name, source_code) in test_files.iter() {
      let stylesheet = create_syle_sheet_parser(source_code);
      let file_name = format_file_name_with_module(file_name, "css_parser", ".css");
      assert_ron_snapshot!(file_name.clone(), stylesheet);
    }
  });
}

#[test]
fn test_html_parser_snapshot() {
  let test_files = read_test_files_with_pattern("tests/golden_tests/html/*.html");
  let settings = setings_snapshot();
  settings.bind(|| {
    for (file_name, source_code) in test_files.iter() {
      let html_root = create_html_parser(source_code);
      let file_name = format_file_name_with_module(file_name, "html_parser", ".html");
      assert_ron_snapshot!(file_name.clone(), html_root);
    }
  });
}
