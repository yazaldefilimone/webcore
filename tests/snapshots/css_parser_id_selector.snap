---
source: tests/snapshot_tests.rs
---
StyleSheet(
  rules: [
    Rule(
      selectors: [
        Simple(SimpleSelector(
          tag_name: None,
          id: Some("main"),
          class: [],
        )),
      ],
      declarations: [
        Declaration(
          name: "color",
          value: ColorValue(HexColorValue("ff0000")),
        ),
      ],
    ),
  ],
)
